### PR TITLE
Fixes #912: Refactor pywbemcli to issue disconnect part2

### LINKED ISSUE #912: Refactor pywbemcli and contextobj classes
Refactor these classes to:
Part 1

1. Eliminate the use of some of the properties in ContextObj such as conn and wbemserver since they are already used in PywbemServer. Specifically this means putting wbem_server and conn in PywbemServer and removing them from ContextObj. Note that this means changing all references to these two attributes which will cause changes to almost all command action methods (i.e. those that use conn or wbemserver) as art of the server connection

2.PywbemServer objects have a connect/disconnect functionality (although the methods are named create_connection() and reset()), which we could use to properly set up and tear down connections. They way we use this functionality currently is not very straight forward, for example "connection select" creates a new context object and globally patches that in, or commands in interactive mode make a deep copy of PywbemServer and modify it, but do not disconnect properly.

Part2

Refactor the pywbemcli code that rebuilds the PywbemServer object in interactive mode. (This will be a separate pr)

### PR DESCRIPTION
**Pull request by KSchopmeyer**
See commit message

NOTE: This PR also removes the temporary PR900 flagged code/comments and actually does the close, etc. and corrects the
issue where the disconnect was not issued under all circmstances of the combination of general options and use of stdin

### CODE REVIEW

File: tests/unit/test_connection_cmds.py [hunk 1]
```diff
@@ -1227,7 +1222,7 @@ class TestSubcmdClass(CLITestsBase):
         "desc, inputs, exp_response, mock, condition",
         TEST_CASES)
     def test_connection(self, desc, inputs, exp_response, mock, condition,
-                        default_connections_file_path):
+                        default_connections_file_path, verbose=True):
```
Comment (andy-maier) [inline tests/unit/test_connection_cmds.py:1225]: The `verbose` argument causes this Pylint issue:
```
************* Module tests.unit.test_connection_cmds
tests/unit/test_connection_cmds.py:1225:55: W0613: Unused argument 'verbose' (unused-argument)
```

**Comment by andy-maier** [comment]
When testing it, I found only one issue, when the connection is specified with -n, -s, or -m in interactive mode:

```
$ echo "-n pegasus server brand" | pywbemcli -v
Connection repository with 5 connection definitions and connections file "/Users/maiera/.pywbemcli_connections.yaml" (exists)
Current connection: "None"
Enter 'help' for help, <CTRL-D> or ':q' to exit pywbemcli or <CTRL-r> to search history, 
Error: No  current server for command "server brand" that requires a WBEM server. Specify a server with the "--server", "--mock-server", or"--name" general option, by setting the corresponding environment variables, or in interactive mode use "connection select" to define a target server
```

The same command in command mode works:
```
$ pywbemcli -v -n pegasus server brand
Connection repository with 5 connection definitions and connections file "/Users/maiera/.pywbemcli_connections.yaml" (exists)
-Debug: PywbemServer connect to https://localhost:5989
OpenPegasus
Debug: PywbemServer disconnect from https://localhost:5989
```

The same command in interactive mode when the connection is specified on the command line also works fine:
```
$ echo "server brand" | pywbemcli -v -n pegasus
Connection repository with 5 connection definitions and connections file "/Users/maiera/.pywbemcli_connections.yaml" (exists)
Enter 'help' for help, <CTRL-D> or ':q' to exit pywbemcli or <CTRL-r> to search history, 
-Debug: PywbemServer connect to https://localhost:5989
OpenPegasus
Debug: PywbemServer disconnect from https://localhost:5989
```

File: pywbemtools/pywbemcli/_pywbem_server.py [hunk 2]
```diff
@@ -443,6 +443,14 @@ def wbem_server(self):
                 verbose=ctx.obj.verbose)
         return self._wbem_server
 
+    @property
+    def connected(self):
+        """
+        Current connected state. Returns True if connected and
+        False if not connected.
+        """
+        return self._wbem_server
```
Comment (andy-maier) [inline pywbemtools/pywbemcli/_pywbem_server.py:452]: This does not return a boolean but the PywbemServer object. Change to using `bool()` around it.

File: pywbemtools/pywbemcli/_pywbem_server.py [hunk 3]
```diff
@@ -519,7 +527,11 @@ def disconnect(self):
 
         Must be connected to the server when calling this method.
         """
-        # print("Debug: PywbemServer disconnect")
+        # TODO(PR900): Remove when disconnect fully tested
+        ctx = click.get_current_context()
+        if ctx.obj:
+            if ctx.obj.verbose:
+                print("Debug: PywbemServer disconnect")
```
Comment (andy-maier) [inline pywbemtools/pywbemcli/_pywbem_server.py:534]: I wonder if we should leave the connect/disconnect messages in, in verbose mode.

**Review by andy-maier — CHANGES_REQUESTED** [review]
See comments

**Comment by andy-maier** [comment]
In addition, the test coverage goes down by 1.4%. This can be seen once the Pylint error is fixed, or in the coverage report of PR #899 : https://coveralls.io/builds/37676122

File: tests/unit/test_connection_cmds.py (hunk 1 above)
Comment (KSchopmeyer) [reply tests/unit/test_connection_cmds.py:1225]: Whoops forgot to remove that one after a test.

File: pywbemtools/pywbemcli/_pywbem_server.py (hunk 3 above)
Comment (KSchopmeyer) [reply pywbemtools/pywbemcli/_pywbem_server.py:534]: Actually I think we should, at least for this release.

File: pywbemtools/pywbemcli/_pywbem_server.py (hunk 2 above)
Comment (andy-maier) [reply pywbemtools/pywbemcli/_pywbem_server.py]: DONE

File: tests/unit/test_connection_cmds.py (hunk 1 above)
Comment (andy-maier) [reply tests/unit/test_connection_cmds.py]: DONE

**Comment by KSchopmeyer** [comment]
Note this corrects the issue Andy found where the 3 forms of  a simple request delivered different results and adds test specifically for that.

**Comment by andy-maier** [comment]
Rebased the PR on the updated PR #915.

**Comment by andy-maier** [comment]
It has gotten better, but is not yet good:

Ok:
```
$ pywbemcli -v -n pegasus server brand
Connection repository with 5 connection definitions and connections file "/Users/maiera/.pywbemcli_connections.yaml" (exists)
-Debug: PywbemServer connect to https://localhost:5989
OpenPegasus
Debug: PywbemServer disconnect from https://localhost:5989
```

Ok:
```
$ echo "server brand" | pywbemcli -v -n pegasus
Connection repository with 5 connection definitions and connections file "/Users/maiera/.pywbemcli_connections.yaml" (exists)
Enter 'help' for help, <CTRL-D> or ':q' to exit pywbemcli or <CTRL-r> to search history, 
-Debug: PywbemServer connect to https://localhost:5989
OpenPegasus
Debug: PywbemServer disconnect from https://localhost:5989
```

Disconnect missing:
```
$ echo "-n pegasus server brand" | pywbemcli -v
Connection repository with 5 connection definitions and connections file "/Users/maiera/.pywbemcli_connections.yaml" (exists)
Current connection: "None"
Enter 'help' for help, <CTRL-D> or ':q' to exit pywbemcli or <CTRL-r> to search history, 
-Debug: PywbemServer connect to https://localhost:5989
OpenPegasus
```

File: pywbemtools/pywbemcli/_pywbem_server.py [hunk 4]
```diff
@@ -519,20 +522,15 @@ def disconnect(self):
...
-            assert self._wbem_server is not None  # Must be connected
-        else:
-            # Old code to be removed once done
-            if self._wbem_server is None:
-                return
+        ctx = click.get_current_context()
+        if ctx.obj:
+            if ctx.obj.verbose:
+                click.echo("Disconnect from server {}".
+                           format(self._mock_server or self._server))
```
Comment (andy-maier) [inline pywbemtools/pywbemcli/_pywbem_server.py:529]: The term "server" is ambiguous here. I suggest to be specific about mock environment vs. WBEM server, e.g. by changing the lines with the click.echo() call to:
```
            if self._mock_server:
                server_txt = "mock environment {}".format(self._mock_server)
            else:
                server_txt = "WBEM server {}".format(self._server)
            click.echo("Disconnecting from {}".format(server_txt))
```

File: pywbemtools/pywbemcli/_pywbem_server.py [hunk 5]
```diff
@@ -551,15 +549,13 @@ def connect(self, log=None, use_pull=None, verbose=None):
           ValueError: Other issues with server-related attributes of this
             object that are considered programming errors.
         """
-        # print("Debug: PywbemServer connect")
+        ctx = click.get_current_context()
+        if ctx.obj:
+            if ctx.obj.verbose:
+                click.echo("Connect to server {}".
+                           format(self._mock_server or self._server))
```
Comment (andy-maier) [inline pywbemtools/pywbemcli/_pywbem_server.py:556]: The term "server" is ambiguous here. I suggest to be specific about mock environment vs. WBEM server, e.g. by changing the lines with the click.echo() call to:
```
            if self._mock_server:
                server_txt = "mock environment {}".format(self._mock_server)
            else:
                server_txt = "WBEM server {}".format(self._server)
            click.echo("Connecting to {}".format(server_txt))
```

File: pywbemtools/pywbemcli/_pywbem_server.py [hunk 6]
```diff
@@ -551,15 +549,13 @@ def connect(self, log=None, use_pull=None, verbose=None):
...
 
-        # TODO(PR900): Remove the PR900 flag test and else path once done.
-        if PR900_REQUIRE_CORRECT_CONNECTION_STATUS:
-            assert self._wbem_server is None  # Must be disconnected
-        else:
-            # Old code to be removed once done
-            if self._wbem_server is not None:
-                return
+        assert self._wbem_server is None  # Must be disconnected
```
Comment (andy-maier) [inline pywbemtools/pywbemcli/_pywbem_server.py:558]: I have some second thoughts on this. If in some more complex scenario we would actually close twice, it is not helpful for a user if we raise the assertion error. This is great during development but not so great during actual usage. Maybe we can introduce an env var that indicates we are running in test.

**Review by andy-maier — CHANGES_REQUESTED** [review]
See comments.

In addition:

* You said there are now test cases that verify the correct connect/disconnect by checking for the verbose messages. I did not find those testcases in this PR.

* The first 3 commits in this PR sum up to no change and can be eliminated, by deleting or squashing them during interactive rebase.

File: tests/unit/test_general_options.py [hunk 7]
```diff
@@ -1307,6 +1313,48 @@ class       Command group for CIM classes.
       'rc': 0,
       'test': 'innows'},
      None, OK],
+
+    #
+    # Test of 3 ways to execute same simple command. Creates a mock
+    # and tests combination of single command,
+    #
```
Comment (andy-maier) [inline tests/unit/test_general_options.py:1320]: These testcases depend on each other, i.e. on the previously generated connections file. That is a general problem, and we should not do that. See also issue #923.

File: pywbemtools/pywbemcli/_pywbem_server.py (hunk 4 above)
Comment (KSchopmeyer) [reply pywbemtools/pywbemcli/_pywbem_server.py:529]: DONE

File: tests/unit/test_general_options.py (hunk 7 above)
Comment (KSchopmeyer) [reply tests/unit/test_general_options.py:1320]: Why not. **DISCUSSION**

Comment (andy-maier) [reply tests/unit/test_general_options.py]: Ok. Let's put this one in, since we have issue #923 for the .bak files.

**Pull request merged**